### PR TITLE
Define and implement an array repetition operator

### PR DESCRIPTION
**Pull request by camh-**
Define the `array * num` operator to repeat `array` `num` number of
times. This allows for easy initialisation of variable sized arrays
without needing to use a loop to do so.

Update the spec with a definition of the operator. Update
`syntax_by_example.md` to give some basic examples.

Implement this in the evaluator and the bytecode compiler/vm.

Issue: https://github.com/evylang/todo/issues/101

### CODE REVIEW

File: docs/spec.md
Comment (juliaogris) [inline docs/spec.md:593]: quote missing after second `"hello`

File: docs/spec.md
Comment (juliaogris) [inline docs/spec.md:596]: how about negative integers?

File: pkg/bytecode/vm.go
Comment (camh-) [inline pkg/bytecode/vm.go:25]: oops. w/while/whole/

File: pkg/bytecode/vm.go
Comment (camh-) [inline pkg/bytecode/vm.go:165]: s/repetitions/len(left.Elements) * repetitions/

File: pkg/bytecode/vm.go
Comment (juliaogris) [inline pkg/bytecode/vm.go:24]: s/integral value/integer value/

**Review by juliaogris — APPROVED** [review]
awesome, thank you so much. a couple of doc nits, otherwise really clear.

**Comment by camh-** [comment]
All fixed and pushed. PTAL

File: pkg/evaluator/evaluator.go
Comment (juliaogris) [inline pkg/evaluator/evaluator.go:734]: Maybe this should also include an error case for negative integers? (and maybe deserves a test?)

File: docs/spec.md
Comment (juliaogris) [inline docs/spec.md:596]: Using a decimal or non-integer value will result in an error (see below)?

**Review by juliaogris — APPROVED** [review]
looks still great to me. just a question around negative repetition counts.

File: pkg/evaluator/evaluator.go
Comment (camh-) [reply pkg/evaluator/evaluator.go:734]: yes it should. good catch. a negative value will crash things I expect

File: docs/spec.md
Comment (camh-) [reply docs/spec.md:596]: what's below? I will change to "non-integer or negative"

File: pkg/evaluator/evaluator.go
Comment (camh-) [reply pkg/evaluator/evaluator.go:734]: done for both evaluator and vm

File: docs/spec.md
Comment (camh-) [reply docs/spec.md]: done. I guess "see below" refers to the comment above? Let me know if you think that reads ok now.

File: pkg/bytecode/compiler.go
Comment (Rossiar) [inline pkg/bytecode/compiler.go:186]:
```suggestion
	if expr.Left.Type().Name == parser.ARRAY && expr.Right.Type() == parser.NUM_TYPE && expr.Op == parser.OP_ASTERISK {
```
being specific about the op doesn't hurt

File: pkg/bytecode/vm.go
Comment (Rossiar) [inline pkg/bytecode/vm.go:24]:
```suggestion
	// repetition operator is not a positive integer value.
```
double negatives are hard to read

File: pkg/bytecode/vm.go
Comment (Rossiar) [inline pkg/bytecode/vm.go:25]: this error message is a bit vague, perhaps something like this?
```suggestion
	ErrBadRepetition = fmt.Errorf("%w: array repetition must be a positive integer", ErrPanic)
```

**Review by Rossiar — COMMENTED** [review]
some small things, overall lgtm

File: pkg/bytecode/vm.go
Comment (camh-) [reply pkg/bytecode/vm.go:24]: Unfortunately the term "positive" excludes zero. "non-negative integer" is the mathematical term to refer to  the integer range 0.. : https://en.wikipedia.org/wiki/Natural_number. I'll try to reword it a different way.

Comment (juliaogris) [reply pkg/bytecode/vm.go:24]: yup, agree, positive is not correct because it excludes zero. The go stdlib uses that term too, e.g.: https://pkg.go.dev/math/rand#Int

File: pkg/bytecode/vm.go
Comment (juliaogris) [reply pkg/bytecode/vm.go:25]: i was also thinking of alternatives for the word count here, but couldn't come up with anything. I also like `array repetition must be a non-negative integer`

Comment (camh-) [reply pkg/bytecode/vm.go:25]: I initially had it more verbose when I had forgotten to check for negative numbers, but simplified it when it got too long. I'll keep this base error the same, but add details where the error is returned.

File: docs/spec.md
Comment (juliaogris) [reply docs/spec.md]: yeah, sorry, the below was below when I made the comment.... you read it correctly.
🍍

Comment (juliaogris) [reply docs/spec.md]: yeah, sorry, the below was below when I made the comment.... you read it correctly.
🍍

**Review by juliaogris — APPROVED** [review]
still 💚

File: docs/spec.md
Comment (camh-) [reply docs/spec.md]: Understood. See above.

File: pkg/bytecode/compiler.go
Comment (camh-) [reply pkg/bytecode/compiler.go]: Thanks. This is arguably a bug. Fixed.

**Comment by camh-** [comment]
do not merge label because I need to squash in the fixup commits before merging. I've forgotten to do that before, so this may help.

**Review by Rossiar — APPROVED** [review]
lgtm

**Pull request merged**
